Fix login issue with pyicloud

PaulCavill opened the pull request:
Upgrade pyicloud version and update field descriptions to reflect that the main account password is required for login.

pyicloud has now been taken over by @timlaing, updated to support apple login flow.


<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
Apple no longer supports app-specific passwords for iCloud. Only your main iCloud account password is supported.

From now on, you’ll need to log in each month and approve the sign-in with a verification code

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
* Update pyicloud version [timlaing/pyicloud@1.0.0...2.0.3](https://github.com/timlaing/pyicloud/compare/1.0.0...2.0.3)


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

Add new requirment pyicloudng module to fix the login flow due to Apple updating login process.

- [x] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue:  #128830
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] I have followed the [perfect PR recommendations][perfect-pr]
- [ ] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

PaulCavill committed:
Move to use pyicloudng from @redcorjo

epenet reviewed (dismissed):
These changes to not match the Home Assistant policy.
The changes need to be implemented in the library: https://github.com/picklepete/pyicloud

If the library is unmaintained, please ensure that every effort is made to contact the author to maybe revive or transfer the ownership of the library to a new maintainer.
If that fails, and you can show that the library cannot be resurected, then you may fork the original library and replace the dependency accordingly.

PaulCavill commented:
Update: New code ready for review, just pending a maintainer action.

https://github.com/picklepete/pyicloud/pull/459

PaulCavill commented:
New maintainer request is underway.
https://github.com/pypi/support/issues/5377

PaulCavill committed:
Fix formatting

PaulCavill committed:
Fix Tabing

PaulCavill committed:
Merge branch 'dev' into task/icloudsrp

PaulCavill commented:
Maybe we can wait another few weeks for the new maintainer to get access.  https://github.com/pypi/support/issues/5377

PaulCavill committed:
Merge branch 'dev' into task/icloudsrp

PaulCavill committed:
Roll some change back getting ready for @timlaing  take over pypi package

PaulCavill committed:
Update string, new version of pyicloud library supports both

PaulCavill committed:
Increase timeout on setup & add optional ReAuth flow

- Increased timeout for setup as Apple's API can be slow and may take longer than 30 seconds to initialize all devices.
- Added an optional config flow to manually trigger ReAuth when the session token expires, as previously only a reboot would initiate reauthentication.

PaulCavill committed:
No need to wait

PaulCavill committed:
Remove unwanted changes

PaulCavill committed:
Merge branch 'dev' into task/icloudsrp

PaulCavill committed:
Merge branch 'dev' into task/icloudsrp

PaulCavill committed:
Merge branch 'dev' into task/icloudsrp

PaulCavill committed:
Remove optional flow

PaulCavill committed:
Remove unwanted changes

PaulCavill committed:
Update pyicloud version, pending release

PaulCavill committed:
Merge branch 'dev' into task/icloudsrp

PaulCavill committed:
Bump to use pyicloud==2.0.1

PaulCavill commented:
New Version of https://pypi.org/project/pyicloud released

PaulCavill committed:
Merge branch 'dev' into task/icloudsrp

joostlek reviewed (dismissed):
Okay so we also need a diff of the changed in the library since previous version and this one. Also, how are existing users affected?

PaulCavill commented on core:
This libary has been takeover by @timlaing as @picklepete was unreachable.

Changes to the library can be found here. https://github.com/timlaing/pyicloud/compare/1.0.0...2.0.1

Release notes: 
https://github.com/timlaing/pyicloud/releases/tag/2.0.0
https://github.com/timlaing/pyicloud/releases/tag/2.0.1

Exiting users will require reauthentication as the saved cookie file name has changed.

joostlek replied:
No, you have `core` added as a submodule

PaulCavill committed:
Merge branch 'dev' into task/icloudsrp

joostlek commented on homeassistant/components/icloud/__init__.py:
```diff
@@ -58,7 +58,7 @@ async def async_setup_entry(hass: HomeAssistant, entry: IcloudConfigEntry) -> bo
         entry,
     )
-    await hass.async_add_executor_job(account.setup)
+    hass.async_add_executor_job(account.setup)
```
But why does it timeout?

PaulCavill committed:
Restore "await"

joostlek commented:
Plaese put the changelog in the PR description

PaulCavill committed:
Update strings to MFA to make clear

PaulCavill committed:
Merge branch 'dev' into task/icloudsrp

joostlek commented on homeassistant/components/icloud/strings.json:
```diff
@@ -6,15 +6,16 @@
         "data": {
           "username": "[%key:common::config_flow::data::email%]",
-          "password": "App-specific password",
+          "password": "Main Password (2fa)",
```
But what happens to existing users? How are they migrated?

joostlek commented:
We're still in discussion, please don't put it ready for review until we resolved those

PaulCavill committed:
Merge branch 'dev' into task/icloudsrp

rcmaehl commented:
Hoping this update can hit soon since iCloud is FUBAR'd otherwise lol

PaulCavill committed:
Merge branch 'dev' into task/icloudsrp

PaulCavill committed:
Merge branch 'dev' into task/icloudsrp

PaulCavill commented:
> [#129059 (comment)](https://github.com/home-assistant/core/pull/129059#discussion_r2151639953) [#129059 (comment)](https://github.com/home-assistant/core/pull/129059#discussion_r2150923504) [#129059 (comment)](https://github.com/home-assistant/core/pull/129059#discussion_r2151638276)
> 
> I still haven't received any answers so far

I have answered all 3 of them.

PaulCavill committed:
Remove core submodule

PaulCavill committed:
Merge branch 'dev' into task/icloudsrp

PaulCavill committed:
Merge branch 'dev' into task/icloudsrp

PaulCavill committed:
Merge branch 'dev' into task/icloudsrp

joostlek committed:
Update homeassistant/components/icloud/strings.json

joostlek commented:
Okay I got good news and bad news, the PR itself looks good now. Except the library depends on a version of fido2 that is not compatible with the `cryptography` version we use. I see that `fido2` has a new version which seems to fix this. So I think @timlaing should look into if `fido2` is required, and if so, make sure it uses a version that works.

Since this PR has been going on for too long I am also not aiming to let it linger around any longer, feel free to send me a message on Discord if you need help with this!

PaulCavill committed:
bump pyIcloud to 2.0.2

PaulCavill committed:
Merge branch 'dev' into task/icloudsrp

PaulCavill commented:
@joostlek pyicloud 2.0.2 has been released to fix the fido issue.

PaulCavill commented:
> Was this even tested at all? I wanted to see if I could help fix the deps issue, but I couldn't even use app passwords to login with pyicloud directly, and it's still an [open issue](https://github.com/timlaing/pyicloud/issues/26) on their project.
> 
> The ["fix"](https://github.com/timlaing/pyicloud/pull/16) for app passwords is from a [commit](https://github.com/icloud-photos-downloader/icloud_photos_downloader/commit/4bcb2ac46a585205cbf3886b3df78179b34b18b1#diff-f2270f557e6afedd3e082f4dd8478d1c96dea051d6ea2e5b1229c0bb58d3d7f4R341) to a 3rd project, icloud_photos_downloader, that apparently didn't make it in any branch, which differs slightly from their [current code](https://github.com/icloud-photos-downloader/icloud_photos_downloader/blob/master/src/pyicloud_ipd/base.py#L359), and that project also has no apparent support for app passwords (there are multiple issues about this closed without fix).
> 
> My understanding is that pyicloud logins like a normal web browser, and you can't use app passwords on the web. The only way it may seem to work is if you somehow end up getting cookies using your main password + 2FA, then change the password, but once the cookies expire and auth is re-triggered it will fail just like before.
> 
> TL;DR, fix need to be tested on a clean slate (removing the icloud dir form HA) or with pyicloud directly.

Apple remove support for App Passwords on web login, so currently the only way to auth is use your main password and MFA each month.

PaulCavill committed:
Merge branch 'dev' into task/icloudsrp

joostlek committed:
Merge branch 'dev' into task/icloudsrp

MartinHjelmare reviewed (changes requested):
Related tests are failing. Please fix them.

PaulCavill commented on homeassistant/components/icloud/manifest.json:
```diff
@@ -6,5 +6,5 @@
   "iot_class": "cloud_polling",
   "loggers": ["keyrings.alt", "pyicloud"],
-  "requirements": ["pyicloud==1.0.0"]
+  "requirements": ["pyicloud==2.0.2"]
```
@MartinHjelmare Review pypi using https://api.github.com/repos/timlaing/pyicloud the details are there....

<img width="491" height="169" alt="image" src="https://github.com/user-attachments/assets/bf62c3b4-5f16-44e2-856a-642633b18be9" />

MartinHjelmare replied:
I don't understand what you're asking.

To be clear, you should update the pyproject.toml file in the library to set the `license` field with a valid license expression with the SPDX license identifier for the library license.

Example:
https://github.com/home-assistant/core/blob/3385151c269febdfcc2a4e43d36b0612500ff535/pyproject.toml#L8

The SPDX identifier for the MIT license is `MIT`.

https://spdx.org/licenses/MIT.html

MartinHjelmare replied:
I don't think so. We don't accept new exceptions for existing libraries. The library should be fixed first.

PaulCavill committed:
Bump pyicloud version

PaulCavill committed:
Merge branch 'dev' into task/icloudsrp

joostlek commented:
Aight, kicked of the CI and updated the PR description, let's see :)

joostlek commented on requirements_test_all.txt:
```diff
@@ -1700,7 +1700,7 @@ pyhomeworks==1.1.2
 
 # homeassistant.components.icloud
-pyicloud==1.0.0
+pyicloud==2.0.2
```
```suggestion
pyicloud==2.0.3
```

joostlek committed:
Apply suggestions from code review

MartinHjelmare commented:
https://github.com/home-assistant/core/pull/129059#pullrequestreview-3180614169

joostlek committed:
fix

joostlek committed:
Merge branch 'task/icloudsrp' of https://github.com/PaulCavill/HAcore into task/icloudsrp

MartinHjelmare reviewed (approved):
Thanks!

MartinHjelmare commented:
We may want to mention in the breaking change release note that the integration requires re-authentication every month.

PaulCavill committed:
Merge branch 'dev' into task/icloudsrp

PaulCavill committed:
Merge branch 'dev' into task/icloudsrp

PaulCavill committed:
Merge branch 'dev' into task/icloudsrp
